src/common/dav/tx.c

```diff
@@ -507,7 +507,25 @@ dav_tx_begin(dav_obj_t *pop, jmp_buf env, ...)
 
 		VALGRIND_START_TX;
 	} else if (tx->stage == DAV_TX_STAGE_NONE) {
```

Comment:
we could consider using something private to the ult.   But it's a good point and bit us in PMDK tx_end before, fixed in this commit to pmdk

`commit 33596644174cc3de6610482fc64b86e6e010059f
Author: Piotr Balcer <piotr@balcer.eu>
Date:   Fri Oct 8 11:18:38 2021 +0200

    obj: make sure nothing is accessed TX_STAGE_NONE cb
    
    The transactional TX_STAGE_NONE callback is executed after
    the end of the transaction indicating that transaction
    can now execute on the same thread. However, currently
    the pmemobj_tx_end() function will actually read from
    the per-thread transaction state after this callback.
    
    This patch addresses the problem by reading all the necessary
    state prior to the callback.`

Reply:
The latest change addresses the following:
- the tx structure elements for the ULT are initialized only after wal_reserve
- the tx structure elements are cleaned up and stage changed to DAV_TX_STAGE_NONE before call to wal_commit

Since the ULT does not yield while in stage other then DAV_TX_STAGE_NONE, I do not see tx getting shared now.

On the call back, the latest fix will cache the callback data before clearing tx and calling wal_commit . Then it calls the callback after wal_commit. This is fine since the callback was already called after stage is set to DAV_TX_STAGE_NONE.